src/server/gc_worker/rawkv_compaction_filter.rs

```diff
@@ -0,0 +1,455 @@
+            self.mvcc_key_prefix.extend_from_slice(mvcc_key_prefix);
+            if commit_ts >= self.safe_point {
+                return Ok(CompactionFilterDecision::Keep);
+            }
+            let raw_value = ApiV2::decode_raw_value(value)?;
+            // the lastest version ,and it's deleted or expaired ttl , need to be send to async gc task
+            if raw_value.is_delete || raw_value.expire_ts.unwrap() < self.current_ts {
+                self.raw_handle_bottommost_delete();
+                if self.mvcc_deletions.len() >= DEFAULT_DELETE_BATCH_COUNT {
+                    self.raw_gc_mvcc_deletions();
+                }
+                // the lastest version ,and it's deleted or expaired ttl , need to be send to async gc task
+            }
+            // the lastest version ,and not deleted or expaired ttl , need to be retained
+            Ok(CompactionFilterDecision::Keep)
+        } else {
+            if commit_ts >= self.safe_point {
+                return Ok(CompactionFilterDecision::Keep);
+            }
+
+            self.filtered += 1;
+            // ts < safepoint ,and it's not the lastest version, it's need to be removed.
+            Ok(CompactionFilterDecision::Remove)
+        }
+    }
+
+    fn raw_gc_mvcc_deletions(&mut self) {
+        if !self.mvcc_deletions.is_empty() {
+            let empty = Vec::with_capacity(DEFAULT_DELETE_BATCH_COUNT);
+            let task = GcTask::RawGcKeys {
+                keys: mem::replace(&mut self.mvcc_deletions, empty), //gc_keys->gc_key->gc->gc.run
+                safe_point: self.safe_point.into(),
+                store_id: self.regions_provider.0,
+                region_info_provider: self.regions_provider.1.clone(),
+            };
+            self.schedule_gc_task(task, false);
+        }
+        self.mvcc_deletions.clear();
+    }
+
+    // `log_on_error` indicates whether to print an error log on scheduling failures.
+    // It's only enabled for `GcTask::OrphanVersions`.
+    fn schedule_gc_task(&self, task: GcTask<RocksEngine>, log_on_error: bool) {
+        match self.gc_scheduler.schedule(task) {
+            Ok(_) => {}
+            Err(e) => {
+                if log_on_error {
+                    error!("compaction filter schedule {} fail", e);
+                }
+                match e {
+                    ScheduleError::Full(_) => {
+                        GC_COMPACTION_FAILURE.with_label_values(&["full"]).inc();
+                    }
+                    ScheduleError::Stopped(_) => {
+                        GC_COMPACTION_FAILURE.with_label_values(&["stopped"]).inc();
+                    }
+                }
+            }
+        }
+    }
+
+    fn raw_handle_bottommost_delete(&mut self) {
+        // Valid MVCC records should begin with `RAW_KEY_PREFIX`.
+        debug!("raw_handle_bottommost_delete:");
+        debug_assert_eq!(self.mvcc_key_prefix[0], RAW_KEY_PREFIX);
+        let key = Key::from_encoded_slice(&self.mvcc_key_prefix);
```

Comment:
You are assuming the field `mvcc_key_prefix` is in encoded format while it seems actually to be raw. You should notice the meaning of the function `from_encoded_slice`.

Reply:
DO NOT store unencoded content in the type `Key`.